documentation/matlab-api.md

```diff
@@ -0,0 +1,381 @@
+Python's main method of handling variable arguments is to use the standard `*args` and `**kwargs` arguments.
+
+```
+def function_name(*args):
+    for arg in args:
+        print(arg)
+
+function_name(1,2,3,4)
+
+def function_name(**kwargs):
+    if kwargs.get('header', False):
+        print(header)
+        del kwargs['header']
+    for arg in kwargs.items():
+        print(arg)
+
+function_name(a=1, header='Hello World', b=4)
+```
+Any Python argument can be treated like a keyword argument.
+```
+def function_name(a,b,c):
+    print(a, b, c)
+
+function_name(1,c=2,b=1) # A is positional
+```
+
+Python also allows default (optional) arguments
+```
+def function_name(a=17):
+    print(a)
+
+function_name(31) # 31
+function_name()   # 17
+```
+
+Any or all of these can be used in combination with each other.
+
+### MATLAB Argument Parsers
+
+### `inputParser`
+Built-in `inputParser` supports positional, optional and required or optional key-value arguments
+
+```
+   defaultHeight = 1;
+   defaultUnits = 'inches';
+   defaultShape = 'rectangle';
+   expectedShapes = {'square','rectangle','parallelogram'};
+
+   p = inputParser;
+   validScalarPosNum = @(x) isnumeric(x) && isscalar(x) && (x > 0);
+   addRequired(p,'width',validScalarPosNum);
+   addOptional(p,'height',defaultHeight,validScalarPosNum);
+   addParameter(p,'units',defaultUnits,@isstring);
+   addParameter(p,'shape',defaultShape,...
+                 @(x) any(validatestring(x,expectedShapes)));
+   parse(p,width,varargin{:});
+```
+
+
+#### Pros
+- Very flexible
+- Well documented with potential online help
+- Standard MATLAB method
+- Explicit argument status
+- Keyword validation
+
+#### Cons
+- May change in future
+- Not extensible if doesn't meet requirements
+
+### `parse_arguments`
+PACE's `parse_arguments` supports positional, required and optional key-value pairs (`'name', value`) and flags (`-flagname`, `-noflagname`, unique substring)
+
+```
+arglist=struct('fix_lattice',0,'fix_alatt',0,'fix_alatt_ratio',0,'fix_angdeg',0,...
+               'fix_orientation',0,'free_alatt',[1,1,1],'free_angdeg',[1,1,1],...
+               'bind_alatt',0);
+flags={'fix_lattice','fix_alatt','fix_alatt_ratio','fix_angdeg','fix_orientation'};
+[args,opt,present] = parse_arguments(varargin,arglist,flags);
+
+(from refine_crystal)
+```
+
+#### Pros
+- Very flexible
+- Well documented
+- Under PACE control
+
+#### Cons
+- Bespoke, requires maintenance
+- Implicit state of arguments
+
+### Raw `varargin` handler
+```
+% Parse input arguments
+if nargin==7
+    expand_qe=false;    % set of distinct q points
+    ...
+    fwhh=varargin{7};
+elseif nargin==5
+    expand_qe=true;     % same q array for each energy in the energy array
+    ...
+    fwhh=varargin{5};
+elseif nargin==4
+    expand_qe=false;
+    ...
+    fwhh=varargin{4};
+else
+    error('Check number of input arguments')
+end
+
+(from disp2sqw)
+```
+
+#### Pros
+- Completely flexible handling of arguments specific to function
+
```

Comment:
PRO: gives you no illusions that you don't have to comment it to understand it. It is quite clear that someone should lay out the input argument options, preferably at this point, and why this degree of complexity is required.

Reply:
I'm not sure one should encourage code that requires comments to be understood as a pro. Comments should be for clarification, not comprehension.